scripts/test-kic-2.x-upgrade.sh

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env bash
+
+# ------------------------------------------------------------------------------
+# test-kick-2.x-upgrade.sh
+#
+# This script is temporary: in the timeframe between 1.x and 2.x KIC
+# releases this script was made to validate that upgrading to the
+# latest pre-release image was functional, once 2.0 releases fully
+# this script will be removed because all tests thereafter will be
+# running on that version and it's lineage.
+# ------------------------------------------------------------------------------
+
+# ------------------------------------------------------------------------------
+# Environment Variables
+# ------------------------------------------------------------------------------
+
+if [[ -z $TEST_ENV_NAME ]]
+then
+    TEST_ENV_NAME="kong-charts-tests"
+fi
+
+# ------------------------------------------------------------------------------
+# Shell Configuration
+# ------------------------------------------------------------------------------
+
+set -euo pipefail
+
+# ------------------------------------------------------------------------------
+# Kubernetes Cluster Setup
+# ------------------------------------------------------------------------------
+
+kubectl cluster-info --context kind-${TEST_ENV_NAME} 1>/dev/null
+KUBERNETES_VERSION="$(kubectl version -o json | jq -r '.serverVersion.gitVersion')"
+
+# ------------------------------------------------------------------------------
+# Determine the latest KIC 2.x Pre-Release Version
+# ------------------------------------------------------------------------------
+
+BASE="kong/kubernetes-ingress-controller"
+LATEST_PRERELEASE=$(curl -s https://api.github.com/repos/${BASE}/releases | \
+    jq -r '.[] | select(.tag_name | test("^2.[0-9]+.[0-9]+")) | .name' | head -1)
+
+if [ "$LATEST_PRERELEASE" = "" ]; then
+    echo "Error: could not find latest release for ${BASE}!${LATEST_PRERELEASE}"
+    exit 1
+fi
+
+echo "INFO: the latest release of ${BASE} v2.x was determined to be ${LATEST_PRERELEASE}"
+
+# ------------------------------------------------------------------------------
+# Setup Chart Cleanup - Kubernetes Ingress Controller
+# ------------------------------------------------------------------------------
+
+RELEASE_NAME="chart-tests-prerelease-compat"
+RELEASE_NAMESPACE="$(uuidgen)"
+
+function cleanup() {
+    echo "INFO: cleaning up helm release ${RELEASE_NAME}"
+    helm delete --namespace ${RELEASE_NAMESPACE} ${RELEASE_NAME}
+    kubectl delete namespace ${RELEASE_NAMESPACE}
+    kubectl delete clusterrole ${RELEASE_NAME}-kong
```

Comment:
This should be a helm post-delete hook instead, no?

Reply:
I don't think so--could you clarify why you think it should? Helm hooks apply to things the chart does when you actually install it, not to stuff specific to test envs.

I think the clusterrole delete here is probably superfluous--`helm delete` should normally purge that since it's a chart-owned resource. The only things you should need to clean up independent of Helm are the namespace and CRDs (which Helm will create, but not delete).

Reply:
These all run as cleanup if the script is interrupted, so if they were in a hook they would not have the same guarantees to execute.

Reply:
@rainest my thinking has been that, if the chart release applies the clusterrole, then it should be responsible for cleaning it up. Passing this responsibility over to the caller (here: the test script) is a sign of a leaky abstraction.

What am I missing?